palm/plugins/dbt/commands/cmd_test.py

```diff
@@ -1,42 +1,99 @@
+@click.option("--select", "-s", multiple=True, help="See dbt docs on select flag")
+@click.option("--exclude", "-e", multiple=True, help="See dbt docs on exclude flag")
+@click.option("--defer", "-d", is_flag=True, help="See dbt docs on defer flag")
+@click.option(
+    "--no-fail-fast", "-nx", is_flag=True, help="Runs all tests even if one fails"
+)
+@click.pass_obj
 def cli(
-    ctx,
+    environment,
     clean: bool,
     no_fail_fast: bool,
+    defer: bool,
     models: Optional[Tuple] = tuple(),
     select: Optional[Tuple] = tuple(),
     exclude: Optional[Tuple] = tuple(),
 ):
-    """Tests the DBT repo"""
+    """Tests the dbt repo"""
+
+    if defer:
+        click.secho("Running 'palm prod-artifacts'...", fg='yellow')
+        exit_code, _, _ = environment.run_on_host("palm prod-artifacts")
+        if exit_code == 2:
+            click.secho(
+                "'palm prod-artifacts' not implemented. Can't use --defer without it!",
+                fg='red',
+            )
+            sys.exit(1)
+        elif exit_code != 0:
+            click.secho(
+                "Something went wrong while pulling the prod artifacts.", fg='red'
+            )
+            sys.exit(1)
+
+    env_vars = set_env_vars(environment, defer)
 
-    cmd = ['dbt', 'test']
-    if select:
+    # --select and --models are interchangeable on dbt >= v1, combine the lists of selections
+    targets = list(set(models + select))
+
+    run_cmd = build_run_command(
+        no_fail_fast=no_fail_fast,
+        targets=targets,
+        exclude=exclude,
+        defer=defer,
+    )
+    success, msg = environment.run_in_docker(run_cmd, env_vars)
+    click.secho(msg, fg="green" if success else "red")
+
+    if clean:
+        success, msg = environment.run_in_docker(
+            "dbt run-operation drop_branch_schemas", env_vars
+        )
+        click.secho(msg, fg="green" if success else "red")
+
+
+def build_run_command(
+    defer: bool = False,
+    no_fail_fast: bool = False,
+    targets: Optional[list] = None,
+    exclude: Optional[Tuple] = None,
+) -> str:
+    cmd = []
+
+    cmd.append(f"dbt test")
+    if targets:
         cmd.append('--select')
-        cmd.extend(select)
-    if models:
-        cmd.append('--models')
-        cmd.extend(models)
+        cmd.extend(targets)
     if exclude:
         cmd.append('--exclude')
         cmd.extend(exclude)
     if not no_fail_fast:
         cmd.append('--fail-fast')
-    if clean:
-        cmd.append('&& dbt run-operation drop_branch_schemas')
+    if defer:
+        if not targets:
+            cmd.extend(["--select", "state:new", "state:modified+"])
+        cmd.append("--defer")
 
-    env_vars = dbt_env_vars(ctx.obj.palm.branch)
-    success, msg = ctx.obj.run_in_docker(" ".join(cmd), env_vars)
-    click.secho(msg, fg="green" if success else "red")
+    return " ".join(cmd)
+
+
+def set_env_vars(environment, defer: bool = False) -> dict:
```

Comment:
I wonder whether we want to extend `dbt_env_vars` in `palm/plugins/dbt/dbt_palm_utils.py` to handle this, rather than defining this additional logic in the command files?

Reply:
Great point. Not sure though, because this file has one set of logic determining these env vars while `palm run` has two (deferral and "iterative" mode) and they don't perform the exact same way.

Reply:
The logic evaluates the same way, though... given `stateful` will be true if `defer` is true. I think we'd want the logic from `palm run` because it allows more control.

Not something we need to do here, but worth thinking about.